src/passes/function_passes/CheckPointPass.cpp

```diff
@@ -0,0 +1,157 @@
+#include <vector>
+#include <map>
+#include <stdint.h>
+#include <llvm/Pass.h>
+#include <llvm/IR/Intrinsics.h>
+#include <llvm/IR/IRBuilder.h>
+#include <llvm/IR/BasicBlock.h>
+#include <llvm/IR/Function.h>
+#include <llvm/IR/Module.h>
+#include <llvm/IR/DataLayout.h>
+#include <llvm/IR/Instruction.h>
+#include <llvm/IR/LegacyPassManager.h>
+#include <llvm/IR/Constants.h>
+#include <llvm/IR/Type.h>
+#include <llvm/Support/raw_ostream.h>
+#include <llvm/Transforms/IPO/PassManagerBuilder.h>
+#define TRACE_FUN_NAME "trace"
+#define GUARD_FUN_NAME "__guard_failure"
+#define UNOPT_PREFIX "__unopt_"
+
+using namespace llvm;
+
+namespace {
+
+  bool isJump(unsigned);
+
+  /*
+   * Insert a stackmap call before each call to `putchar`.
```

Comment:
Let's add some issues on this repo for things to look at if we get time.

The first, I think is that insertion of stackmap calls should be general, right?

Reply:
I'm working on addressing this issue (it isn't going to be part of this PR though). In some sense, it should be general. However, each guard failure needs to trigger a `patchpoint` call (`patchpoint` calls transfer control to my runtime). In turn, each `patchpoint` call in the optimized trace needs to have an associated `stackmap` call in the unoptimized version.